### ISSUE TITLE
dart: Cannot type DOM accesses

### ISSUE THREAD

**Issue by alexeagle** [angular/angular#2770]
I'm adding return types as a first pass of adding a linter for better API completeness.

Many methods in the DOM interfaces, eg. `*_dom_strategy.ts`, `dom_adapter.ts`, cannot be typed because the dart_analyzer is configured to know that neither parse5 nor a browser might be present at runtime. However, leaving all these typed as `any` weakens our ability to use these APIs and get typical benefits of typing.

I'll leave the proper types commented out with a reference to this issue.

Related: @mprobst has a proposal to fix this (IIUC) with "DOM abstraction layer for Angular" https://docs.google.com/document/d/1VO8L2ZyVKTCGsO_JGZYwii3Owu5Lz0qwG35ApPL-TcY/edit#heading=h.66i8qgknkokg
Martin, please close as a dupe if there is already an issue tracking that.

**Comment by alexeagle** [comment]
I would like to type these methods for the benefit of TypeScript usage. Here is a proposal: add a syntax clue for ts2dart which says the type should be replaced by `dynamic` even if it's not any. I imagine doing it with a comment, but maybe there is a different syntax bit that could work?
@yjbanov what do you think?
Otherwise I am replacing them with a comment pointing at this issue, eg. `List</*(#2770) Node*/any>`

**Comment by alexeagle** [comment]
Yegor had a simple suggestion: add `Node`, etc (maybe 5 or 6 DOM types) to `TS_TO_DART_TYPENAMES` in ts2dart, mapped to dynamic.
I need to verify whether there are other places where we want these types to have the present identity mapping, because the appropriate DOM api is actually present to dart_analyzer.

**Comment by mprobst** [comment]
Adding such fake type mappings SGTM, though in that case we should make the type mappings smarter - currently they are simple string substitutions and don't look at the declaration location of the symbol. `Node`, `Element`, and so on are very generic type names, so we should improve our type name substitution to match on both, i.e. `(declaration location, symbol name)` instead of just `symbol name`.

**Comment by mprobst** [comment]
Also, just to be clear, the DOM abstraction proposal you linked to above turned out not to work with `ts2dart`.

**Comment by alexeagle** [comment]
Sounds good - I'll have to wire the TypeChecker to this location so it will need a bit of refactoring.

**Comment by yjbanov** [comment]
Just so we're on the same page. This can only be a _temporary_ solution. We still need to find a long-term solution that does not blindly erase types in the resulting Dart code. As an example, everything under `modules/angular2/src/render/dom` is currently untyped for no good reason (see e.g. DomElement, DomView, LightDom).

Do we have an issue already or should I log one?

**Comment by alexeagle** [comment]
Let's keep this issue open for the long-term solution (because Dart is in the title)
I don't have an idea what to do for that, though. Assign to you for now.

**Comment by mprobst** [comment]
@yjbanov I've added you to a thread with the Dart people on this - we're blocked by Dart making progress here. The `dart:html` and `package:html` packages are incompatible, there simply is no way to express 'an Element, either from dart:html or package:html' in Dart.

**Comment by mprobst** [comment]
@naomiblack @yjbanov how urgent is this? Does this immediately impact users?

This is a language design level issue in Dart currently, and the design seems to be unclear still, so I expect a fix to be multiple months away (AFAICT).

There is the simpler solution of just having common super-interfaces for the two libraries (which has other issues, but still...), so we might be able to expedite.

**Comment by naomiblack** [comment]
I am moving this to the backlog -- we should continue to track but I think the immediate work is on hold. Should the PR be closed?

**Comment by naomiblack** [comment]
Closing, obsolete. Not clear this is an urgent issue for GT.

**Comment by zoechi** [comment]
Configurable imports seems to be whats required here and that's about to be shipped.
